text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,183 @@
+
+- Bootstrapping trust: how does the registry decide to trust a new user?
+- Key generation: where and how the key pair is made?
+- Key rotation: how often do existing users need to make a new key pair?
+- Revocation: how does the registry decide to stop trusting an existing key pair?
+
+In order for crates.io to support asymmetric tokens these questions will need to be answered for crates.io. If and how crates.io will implement compatibility with these new tokens will be left for a follow-up discussion/RFC.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Private registries that require authentication use asymmetric cryptography as a more secure way for cargo to log in. Each registry works a little different, but the most common workflow is:
+1. Generate a key pair. (For many registries, you can generate the key pair using the cargo command `cargo login --generate-keypair`, which will print the public key for use in the next step.)
+2. Go to the registries log in page, upload your public key and get the user ID for that key pair.
+3. On the command line run `cargo login --registry=name --private-key-path=path userId`
+
+There are credential processes for using key pairs stored on hardware tokens. Check crates.io to see if there's one available for your hardware. Each one is a little different, but the general workflow is:
+1. `cargo install credential-process-for-your-hardware-token`
+2. run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
+3. Go to the registries log in page, upload your public key and get your user ID.
+4. Edit `credentials.toml` to have a `credential-process` field as described by `credential-process-for-your-hardware-token` docs. (The credential process command may help do this for you.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In [`config.toml`](https://doc.rust-lang.org/cargo/reference/config.html) and `credentials.toml` files there is a field called `private-key-path`, which provides a path to a `PKCS#12` formatted file containing a private key.
+
+A keypair can be generated with `cargo login --generate-keypair` which will:
+- generate a public/private keypair in the currently recommended fashion.
+- save the private key in the default location.
+- print the public key and the path to the file.
+  (See unresolved questions section.)
+
+There is also a field called `user-id` which is a string chosen by the registry, which is intended to be non-secret and used for identifying the user. Cargo requires it to be non-whitespace printable ASCII, registries should base64 encode non-ASCII data. Cargo may broaden the allowed set of characters to some other subset of UTF-8 in the future.
```

Comment:
I'm not sure a `user-id` is a good idea. Especially as people change can change usernames and sometimes want their old names gone from the records, so ideally one designs technology that does without usernames. I guess the purpose of the `user-id` is to figure out which public key the private key corresponds to? For this purpose, sending a *key* identifier would make more sense, as cargo supports multiple tokens per user. The key identifier would be purely derived from the key so the user wouldn't even have to enter it, which reduces setup overhead.

Reply:
I really like the idea of using a "key ID", say the SHA-384 of the public key! This removes a step for most use cases. For the rare use cases where the server does want some arbitrary information to be passed back as part of the token we can add a field for that purpose. (Here is cryptographic proof that the central CA server authorized this action. Or whatever this particular registry needs.) That field can have a less confusing name. I did not have time to edit this into the RFC today, but I will do so and my next opportunity.

Reply:
A standard way of encoding that information is via [SPKI key fingerprints](https://docs.rs/spki/latest/spki/struct.SubjectPublicKeyInfo.html)